connector/jobrunner/channels.py

```diff
@@ -0,0 +1,602 @@
+        if o in self._known:
+            return
+        self._known.add(o)
+        heappush(self._heap, o)
+
+    def remove(self, o):
+        if o is None:
+            raise ValueError()
+        if o not in self._known:
+            return
+        if o not in self._removed:
+            self._removed.add(o)
+
+    def pop(self):
+        while True:
+            try:
+                o = heappop(self._heap)
+            except IndexError:
+                # queue is empty
+                return None
+            self._known.remove(o)
+            if o in self._removed:
+                self._removed.remove(o)
+            else:
+                return o
+
+
+class SafeSet(set):
+    """A set that does not raise KeyError when removing non-existent items.
+
+    >>> s = SafeSet()
+    >>> s.remove(1)
+    >>> len(s)
+    0
+    >>> s.remove(1)
+    """
+    def remove(self, o):
+        try:
+            super(SafeSet, self).remove(o)
+        except KeyError:
+            pass
+
+
+class ChannelJob:
+    """A channel job is attached to a channel and holds the properties of a
+    job that are necessary to prioritise them.
+
+    Channel jobs are comparable according to the following rules:
+        * jobs with an eta come before all other jobs
+        * then jobs with a smaller eta come first
+        * then jobs with smaller priority come first
+        * then jobs with a smaller creation time come first
+        * then jobs with a samller sequence come first
+
+    Here are some examples.
+
+    j1 comes before j2 before it has a smaller date_created
+    >>> j1 = ChannelJob(None, 1, seq=0, date_created=1, priority=9, eta=None)
+    >>> j1
+    <ChannelJob 1>
+    >>> j2 = ChannelJob(None, 2, seq=0, date_created=2, priority=9, eta=None)
+    >>> j1 < j2
+    True
+
+    j3 comes first because it has lower priority,
+    despite having a creation date after j1 and j2
+    >>> j3 = ChannelJob(None, 3, seq=0, date_created=3, priority=2, eta=None)
+    >>> j3 < j1
+    True
+
+    j4 and j5 comes even before j3, because they have an eta
+    >>> j4 = ChannelJob(None, 4, seq=0, date_created=4, priority=9, eta=9)
+    >>> j5 = ChannelJob(None, 5, seq=0, date_created=5, priority=9, eta=9)
+    >>> j4 < j5 < j3
+    True
+
+    j6 has same date_created and priority as j5 but a smaller eta
+    >>> j6 = ChannelJob(None, 6, seq=0, date_created=5, priority=9, eta=2)
+    >>> j6 < j4 < j5
+    True
+
+    Here is the complete suite:
+    >>> j6 < j4 < j5 < j3 < j1 < j2
+    True
+
+    j0 has the same properties as j1 but they are not considered
+    equal as they are different instances
+    >>> j0 = ChannelJob(None, 1, seq=0, date_created=1, priority=9, eta=None)
+    >>> j0 == j1
+    False
+    >>> j0 == j0
+    True
+    """
+
+    def __init__(self, db_name, channel, uuid,
+                 seq, date_created, priority, eta):
+        self.db_name = db_name
+        self.channel = channel
+        self.uuid = uuid
+        self.seq = seq
```

Comment:
It's my idea to implement guaranteed sequence on job execution. date_created is not sufficient, and priority is another concept.

It is not fully implemented yet, because jobs that are retried get an eta and this does not work well with sequentiality... needs further thinking.